### PR TITLE
save fixtures to a separate database

### PR DESCRIPTION
**Pull request by czue**
depends on https://github.com/dimagi/commcare-hq/pull/5385
# Fixtures Migration
## Replicate fixtures to new database

The easiest way to do this is the following:

```
./manage.py couch_migrate corehq/apps/cleanup/management/commands/couch_migrations/fixtures.json --replicate
```

Behind the scenes this is the equivalent of creating a config file like this:

```
{
    "source": "http://localhost:5984/hqdev",
    "target": "http://localhost:5984/hqdev__fixtures",
    "continuous": true,
    "filter": "hqadmin/doc_types",
    "query_params": {
        "doc_types": "FixtureDataType FixtureDataItem FixtureOwnership"
    }
}
```

and running a replication command like this:

```
curl -H 'Content-Type: application/json' -X POST http://localhost:5984/_replicate -d @replicate-fixtures.json
```
## Check views for consistency

```
./manage.py couch_migrate corehq/apps/cleanup/management/commands/couch_migrations/fixtures.json --check
```
## Test

Things I've tested
- Table list
- Viewing a single table
- Downloading tables
- Uploading tables
- OTA restore

### CODE REVIEW

**Comment by czue** [comment]
@dannyroberts @snopoke fyi

**Comment by dimagimon** [comment]
There are 12 errors:

* corehq/apps/cleanup/management/commands/couch_migrate.py, line 2 - F401 'Database' imported but unused
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 3 - F401 'settings' imported but unused
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 17 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 18 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 19 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 20 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 22 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 23 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 24 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 25 - E127 continuation line over-indented for visual indent
* corehq/apps/cleanup/management/commands/couch_migrate.py, line 78 - W391 blank line at end of file
* corehq/apps/hqadmin/_design/filters/doc_types.js, line 1 - Missing name in function declaration.

File: corehq/apps/hqadmin/_design/filters/doc_types.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+function (doc, req) {
```
Comment (dimagimon) [inline corehq/apps/hqadmin/_design/filters/doc_types.js]: Missing name in function declaration.

File: corehq/apps/hqadmin/_design/filters/doc_types.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    var doc_types_string = req.query.doc_types;
+    if (doc_types_string) {
+        var doc_types = doc_types_string.split(' ');
+        if (doc_types.indexOf(doc.doc_type) >= 0) {
+            return true;
+        }
+    }
+    return false;
+}
```
Comment (snopoke) [inline corehq/apps/hqadmin/_design/filters/doc_types.js]: Similar to https://github.com/dimagi/fluff/blob/master/fluff/fluff_filter/_design/filters/domain_type.js but this supports multiple doc types

**Comment by snopoke** [comment]
nice

File: corehq/apps/hqadmin/_design/filters/doc_types.js (hunk 2 above)
Comment (czue) [inline corehq/apps/hqadmin/_design/filters/doc_types.js]: ah, i was actually looking for this. do you think it's worth changing and removing this one?

Comment (snopoke) [inline corehq/apps/hqadmin/_design/filters/doc_types.js]: I think if you're going to add a new filter its probably worth including the domains part as well and then we can just kill the fluff one later

File: corehq/apps/hqadmin/_design/filters/doc_types.js (hunk 1 above)
Comment (dimagimon) [inline corehq/apps/hqadmin/_design/filters/doc_types.js]: Missing name in function declaration.

**Comment by dannyroberts** [comment]
Really nice solution! Impressed with the simplicity of what was otherwise a huge headache :) Automation ftw

File: corehq/apps/hqadmin/_design/filters/doc_types.js (hunk 2 above)
Comment (czue) [inline corehq/apps/hqadmin/_design/filters/doc_types.js]: https://github.com/dimagi/commcare-hq/pull/5385

File: corehq/apps/hqadmin/_design/filters/domains_and_doc_types.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+function (doc, req) {
```
Comment (dimagimon) [inline corehq/apps/hqadmin/_design/filters/domains_and_doc_types.js]: Missing name in function declaration.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dannyroberts** [comment]
I may have missed an off-thread discussion, but are we poised to do this as soon as our cloudant migration is complete? I imagine that given the recent situations that we should have a plan for easing into and backing out of this plan if we see it causing problems; but overall I'm pretty excited to see this happen and see what other models we can apply it to.

**Comment by czue** [comment]
There are problems with the replication that have prevented it from being doable now. Definitely want to get it out as soon as those are resolved.

**Comment by snopoke** [comment]
Any update on this>?

**Comment by czue** [comment]
I was hoping that the cloudant migration would allow the replication to go though, but no luck. i'm going to write a command to manually sync them across. probably won't get to it until after interrupt week, sadly

**Comment by snopoke** [comment]
Is it just this replication that isn't working? Can't you push it to
Cloudant to sort out?

**Comment by czue** [comment]
That hasn't been going well so far. This was the last correspondence:

Apologies for the delay.  Our operators have taken a look into this issue today, unfortunately there is not much more we can do to help the replication complete.  The problem is from a '_changes' request timing out which looks to be caused by large documents/attachments.  Our operators attempted reconfiguring the cluster to make the data pass through the filter (unfortunately, accidentally causing those 500's you saw earlier) but there is not much else that can be done.  Large documents and filtered replication can be problematic even by themselves, this is a combination of both.

They suggested just doing it manually. I think they're willing to help, but it's probably easier and less annoying long term to just have that capability ourselves.

**Comment by snopoke** [comment]
:poo:

On 15 April 2015 at 17:38, Cory Zue <notifications@github.com> wrote:

> That hasn't been going well so far. This was the last correspondence:
>
> Apologies for the delay. Our operators have taken a look into this issue
> today, unfortunately there is not much more we can do to help the
> replication complete. The problem is from a '_changes' request timing out
> which looks to be caused by large documents/attachments. Our operators
> attempted reconfiguring the cluster to make the data pass through the
> filter (unfortunately, accidentally causing those 500's you saw earlier)
> but there is not much else that can be done. Large documents and filtered
> replication can be problematic even by themselves, this is a combination of
> both.
>
> They suggested just doing it manually. I think they're willing to help,
> but it's probably easier and less annoying long term to just have that
> capability ourselves.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5372#issuecomment-93453620>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/app_manager/tests/test_views.py, line 14 - F401 'CareplanModule' imported but unused
* corehq/apps/cloudcare/tests/test_dbaccessors.py, line 24 - F812 list comprehension redefines 'o' from line 19
* corehq/apps/facilities/tests.py, line 17 - F812 list comprehension redefines 'o' from line 13
* corehq/apps/facilities/tests.py, line 24 - F812 list comprehension redefines 'o' from line 20
* corehq/apps/hqmedia/models.py, line 540 - F402 import '_' from line 18 shadowed by loop variable

**Comment by czue** [comment]
did someone just merge this with master or something?

this should literally be a one line change. I can just rebase and force push though.

**Comment by czue** [comment]
(nevermind, just saw https://github.com/dimagi/commcare-hq/pull/7232#issuecomment-115697471)

**Comment by dannyroberts** [comment]
Yeah, and actually, feel free to rebase—now that `rewrite-trivial-views` is merged, that should clean up the history a bit

**Comment by dimagimon** [comment]
There are 21 errors:

* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 114 - E501 line too long (154 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 120 - E501 line too long (124 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 125 - E121 continuation line under-indented for hanging indent
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 125 - E501 line too long (129 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 126 - E501 line too long (125 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 129 - E121 continuation line under-indented for hanging indent
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 136 - E121 continuation line under-indented for hanging indent
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 141 - E501 line too long (160 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 148 - E501 line too long (176 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 150 - E501 line too long (140 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 155 - E501 line too long (129 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 156 - E501 line too long (128 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 172 - E121 continuation line under-indented for hanging indent
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 249 - E501 line too long (138 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 256 - E501 line too long (144 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 266 - E501 line too long (144 > 115 characters)
* corehq/apps/app_manager/tests/test_bulk_app_translation.py, line 268 - E501 line too long (134 > 115 characters)
* corehq/apps/cloudcare/tests/test_dbaccessors.py, line 24 - F812 list comprehension redefines 'o' from line 19
* corehq/apps/facilities/tests.py, line 17 - F812 list comprehension redefines 'o' from line 13
* corehq/apps/facilities/tests.py, line 24 - F812 list comprehension redefines 'o' from line 20
* corehq/apps/fixtures/fixturegenerators.py, line 56 - F841 local variable '_' is assigned to but never used

File: corehq/apps/fixtures/_design/views/data_types_by_domain/map.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+function (doc) {
```
Comment (dimagimon) [inline corehq/apps/fixtures/_design/views/data_types_by_domain/map.js]: Missing name in function declaration.

Comment (dimagimon) [inline corehq/apps/fixtures/_design/views/data_types_by_domain/map.js]: Missing name in function declaration.

**Comment by benrudolph** [comment]
(broke staging build so merged master)

**Comment by czue** [comment]
@benrudolph I think we can pull this off staging.

**Comment by benrudolph** [comment]
👍

**Comment by dannyroberts** [comment]
Closing this in favor of doc type migrations (which upon a little sanity checking still seems like it's going to be pretty straightforward)

**Comment by czue** [comment]
:+1: makes sense. this PR was my white whale anyways...

**Comment by dannyroberts** [comment]
:whale:

**Pull request closed without merging**